shoop/core/cache.py

```diff
@@ -0,0 +1,161 @@
+# -*- coding: utf-8 -*-
+"""
+Utilities for caching, with versioned caching and automatic timeout determination.
+
+Versioning works by way of "namespaces". Namespaces are the first colon-separated part
+of cache keys.
+
+For instance, the cache keys `price:10`, `price:20`, and `price` all belong to the
+`price` namespace and can be invalidated with one `bump_version("price")` call.
+
+The versions themselves are stored within the cache, within the `_version` namespace.
+(As an implementation detail, this allows one to invalidate _all_ versioned keys by
+bumping the version of `_version`. Very meta!)
+"""
+import time
+from django.conf import settings
+from django.utils.encoding import force_text
+from django.core.cache import caches
+from django.core.signals import request_finished
+from threading import local
+
+DEFAULT_CACHE_DURATIONS = {}
+_versions = local()
+
+
+def _clear_versions_for_request(**kwargs):
+    _versions.__dict__.clear()
+
+
+request_finished.connect(_clear_versions_for_request, dispatch_uid="shoop.core.cache._clear_versions_for_request")
+
+
+def _get_cache_key_namespace(cache_key):
+    """
+    Split the given cache key by the first colon to get a namespace.
+
+    :param cache_key: Cache key string
+    :type cache_key: str
+    :return: Cache namespace string
+    :rtype: str
+    """
+    return force_text(cache_key).split(":")[0]
+
+
+def get_cache_duration(cache_key):
+    """
+    Get a suitable cache duration for the given cache key (by the default duration and namespace mapping).
+
+    :param cache_key: Cache key string
+    :type cache_key: str
+    :return: Timeout seconds
+    :rtype: int
+    """
+    namespace = _get_cache_key_namespace(cache_key)
+    duration = settings.SHOOP_CACHE_DURATIONS.get(namespace)
+    if duration is None:
+        duration = DEFAULT_CACHE_DURATIONS.get(namespace, settings.SHOOP_DEFAULT_CACHE_DURATION)
+    return duration
+
+
+class VersionedCache(object):
+    def __init__(self, using):
+        """
+        :param using: Cache alias
+        :type using: str
+        """
+        self.using = using
+        self._cache = caches[self.using]
+
+    def bump_version(self, cache_key):
+        """
+        Bump up the cache version for the given cache key/namespace.
+
+        :param cache_key: Cache key or namespace
+        :type cache_key: str
+        """
+        namespace = _get_cache_key_namespace(cache_key)
+        version = str(time.time())
```

Comment:
This could also have the `assert isinstance(cache_key, six.text_type)` and the version should be formed with `"%s" % time.time()` with unicode_literals active so it will be unicode on both py2 and py3.